README.md

```diff
@@ -1,25 +1,73 @@
+
+### Garbage collection
+
+This plugin runs periodic garbage collection to delete `sdv_fanout_data` record if there is no reference left from `sdv_fanout_ids` table.
+
+The `sdv_fanout_data` and `sdv_fanout_ids` table will be deleted if the timestamp is older than the configured retention period.
+
+## Data Flow
+
+This plugin starts a pool of dispatcher processes to fanout the data to the locally connected subscribers.
+
+Below are the events which will trigger the fanout to the subscribers:
+
+### Dispatch Triggers
+
+- **Batch Received**:
+  When a new batch is received, the publishing client will trigger the fanout data (and IDs) to be written to the database, then send notification to the dispatcher pool.  Depending on the session lookup result (pid), the notification `{maybe_send, new_batch, Pid, VIN}` should be sent to a dispatcher in the pool or an `rpc` call to a remote node to do the same.
+- **Heartbeat**:
+  After connected, the vehicle publishes to topic `ecp/${VIN}/online`, then it periodically (every 1m) publishes to the topic `ecp/${VIN}/heartbeat`. The plugin implemented callback (`'message.publish'`) should trigger a `{maybe_send, heartbeat, Pid, VIN}` notification to the dispatcher pool. This is to ensure messages are sent to the vehicle even if other triggers missed due to race conditions. For example, after client session resume it may not observe the new batch being inserted in another node, and the new batch handler in the other node may not see the session being online, hence both nodes miss the opportunity to send the message. This heartbeat is to ensure the message is eventually sent.
+- **PUBACK Received**:
+  When a `PUBACK` is received, the plugin implemented callback (`'delivery.completed'`) should delete the message ID from the `sdv_fanout_ids` table, and trigger a `{acknowledge, self(), RequestID}` notification to the dispatcher pool.
+
+### Dispatch Process
+
+Each dispatcher process maintains a inflight state of the messages to be sent to the subscribers.
+
+The inflight state is a ETS table of tuples of `{Key = SubPid, VIN, Ts, RequestID}`.
+
+The dispatcher should call `emqx_cm:is_channel_connected/1` to check if the subscriber is currently online. Ignore the notification if the subscriber is not online.
+
+If there is already a message in flight, the `maybe_send` notification will be ignored.
+
+If there is no message in flight, the dispatcher will read the `sdv_fanout_ids` table to check if there is any message to be sent to the subscriber, if found, the dispatcher will send the message to the subscriber with QoS=1 topic = `agent/${VIN}/proxy/request/${request_id}`, monitor the subscriber process, and insert the sent message into the inflight table.
+
+If a `'DOWN'` message is received from the subscriber, the dispatcher will remove the subscriber from the inflight table. The next `maybe_send` notification will be sent to the dispatcher pool after the vehicle reconnects and send heartbeat again.
+
+## TODO
```

Comment:
I think SDV platform could split the message into two, one message to provide the `data`, another message for distribution list then it does not need to upload data every time and EMQX does not need to run sha1 on that.

Reply:
I will discuss with the team

Reply:
agreed.
will change to two topics
`$SDV-FANOUT/data` for payload (identify by its sha1)
`$SDV-FANOUT/trigger` for a batch of notification (`#{ids := VINs, request_id := RequestID, data_id := Sha1}`)